src/names.md

```diff
@@ -0,0 +1,143 @@
+# Names
+
+An *entity* is a language construct that can be referred to in some way within
+the source program, usually via a [*path*]. Entities include [types], [items],
+[generic parameters], [variable bindings], [loop labels], [lifetimes],
+[fields], [attributes], and [lints].
+
+A *declaration* is a syntactical construct that can introduce a *name* to
+refer to an entity. Entity names are valid within a [*scope*] — a region of
+source text where that name may be referenced.
+
+Some entities are [explicitly declared](#explicit-entities) in the source
+code, and some are [implicitly declared](#implicit-entities) as part of the
+language or compiler extensions.
+
+[*Paths*] are used to refer to an entity, possibly in another scope. Lifetimes
+and loop labels use a [dedicated syntax][lifetimes-and-loop-labels] using a
+leading quote.
+
+Names are segregated into different [*namespaces*], allowing entities in
+different namespaces to share the same name without conflict.
+
+[*Name resolution*] is the compile-time process of tying paths, identifiers,
+and labels to entity declarations.
+
+Access to certain names may be restricted based on their [*visibility*].
+
+## Explicit entities
```

Comment:
Can we change these section titles to "Explicitly declared entities" and "Implicitly declared entities"?